### PR TITLE
Implement all effects

### LINKED ISSUE #1: `take(channel, pattern)` support
Hey, I'm working on a PR, but I wanted to run this by you first; the Take effect also has a variant that works with channels. It's fairly trivial to add this to typed-redux-saga's implementation, but since typescript doesn't support generator overloading I had to change it to a function with an IIGE (immediately invoked generator expression :).
```ts
export function take<A extends Action>(
  channel: Channel<A>,
  pattern: ActionPattern<A>,
): SagaIterator<A>;
export function take<A extends Action>(
  pattern?: ActionPattern<A>,
): SagaIterator<A>;
export function take<A extends Action>(a?: any, b?: any): SagaIterator<A> {
  return (function*() {
    if (a.take) {
      return yield rawTake(a, b);
    } else {
      return yield rawTake(b);
    }
  })();
}
```
This changes the type of the export, but not the signature. It's also possible to keep the generator and avoid the overload, but it's really ugly. Are you cool with using a function for the top level export, and if so, should the rest of the exports use the same pattern? In practice, it doesn't change the developer ergonomics re: intellisense.

### LINKED ISSUE #3: Port from tslint to eslint

### LINKED ISSUE #8: `getContext()` support
is there any possibility of adding support for the `getContext()` effect?

### PR DESCRIPTION
**Pull request by gilbsgilbs**
**Changelog:**

- Implement and provide types for all effects from redux-saga
- Ported the actual library code from TS to JS (see explanation below)
- As a consequence, build the project using babel instead of tsc
- Setup ESLint
- Add typing tests and linting through dtslint

Fixes #1  
Fixes #3  
Fixes #8  

Note that this is a ~major~ complete rewrite of the whole library and I would completely understand should you say to me that I should maintain my own fork.

As to why I migrated the code from TS to JS: TS was complaining about conflicting method definitions. Making the code in JS is actually simpler, but if you know how to fix this I'm happy to revise.

I'm not sure how far we could get regarding automation. I think the JS code generation is "good-enough", but the typedefs are just a wild copy-paste from redux-saga's typedefs. I just changed the return value of every effect to the `SagaGenerator` stuff and tried to ensure all the return types were correct. However, some return types ended up being quite complicated (like the `all` and `race` effects).

Finally, unless I'm missing something regarding automation, I don't see this getting merged into redux-saga's codebase anytime soon unfortunately. IMHO, the only acceptable way would be to drop support for `yield` effect in redux-saga, which is a huge breaking change to satisfy only a few typescript users. It's unlikely the community would accept this. :disappointed: Maybe the maintenance would be reduced if we just did a fork of redux-saga. This way the only job of the mainteners would be to merge upstream from time to time, updating type definitions as they change.

Before I forget:
- [ ] Task type from Redux-Saga could be improved (see `FIXME` in `index.d.ts`)
- [ ] I feel like it's unconvenient for the `race` and `all` `yield*` effects to take raw redux-saga's effects. Is it possible and/or desirable to pass `yield*` effects to them? If so, it would be quite easy to add a simple linter rule that bans `redux-saga/import`. I would kind of like that. WDYT?

And now I need some sleep :smile: 

ping @danielnixon @Andarist

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+): SagaGenerator<never, ForkEffect<never>>;
+
+export function retry<Fn extends (...args: any[]) => any>(
+  maxTries: number,
+  delayLength: number,
+  fn: Fn,
+  ...args: Parameters<Fn>
+): SagaGenerator<SagaReturnType<Fn>, CallEffect<SagaReturnType<Fn>>>;
+
+type EffectReturnType<T> = T extends CallEffect
```
Comment (gilbsgilbs) [inline types/index.d.ts]: Any idea how to improve this heuristic?

**Comment by danielnixon** [comment]
Wow, this is great. Thanks @gilbsgilbs.

**Comment by Andarist** [comment]
> I don't see this getting merged into redux-saga's codebase anytime soon unfortunately. 

I'm not opposed to merging this into redux-saga itself - I just don't think we should change this to be a default, but rather an opt-in from TS users, which would merely mean that an alternative entry point would have to be used instead of the default `redux-saga` (smth like `@redux-saga/typed` maybe).

I just don't have time right now to migrate this myself, setup everything etc. I can discuss how I believe this should be done though, code reviewing and maybe do some work around our babel plugin so the overhead from this could be transpiled away.

**Comment by gilbsgilbs** [comment]
Thanks @Andarist .

> I'm not opposed to merging this into redux-saga itself - I just don't think we should change this to be a default, but rather an opt-in from TS users, which would merely mean that an alternative entry point would have to be used instead of the default redux-saga (smth like @redux-saga/typed maybe).

Yeah, I'm definitely in phase with that. The problem is that I don't see how to implement them without duplicating all the types. Considering that [!1892](https://github.com/redux-saga/redux-saga/pull/1892) has already duplicated all the typedefs for TS3.6+, I don't really want to add significantly more maintenance cost to redux-saga's typings unless you're very comfortable with it. Although well, maintenance in this repo or in redux-saga's repo, what difference after all?

> maybe do some work around our babel plugin

I was also thinking that we could provide a babel macro.

File: types/index.d.ts (hunk 1 above)
Comment (danielnixon) [reply types/index.d.ts]: I've updated it in https://github.com/agiledigital/typed-redux-saga/commit/2aebeb1ccc464280986e6a4ed253ad5955ed4f10#diff-88bec6beae84369c0376b56b3bb88fe1R468 to include `SagaGenerator`

**Comment by jmrossy** [comment]
> I'm not opposed to merging this into redux-saga itself 
@Andarist can you elaborate on what's the downside of this being the default? Would the main issue be the requirement to use `yield*` instead of yield?

**Pull request merged**